// routers/lib/error-handler.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _invariant = require('invariant');

var _invariant2 = _interopRequireDefault(_invariant);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var handlers = {
  boom: function boom(err) {
    if (err.isBoom) {
      var _err$output$payload = err.output.payload;
      var statusCode = _err$output$payload.statusCode;
      var error = _err$output$payload.error;
      var message = _err$output$payload.message;


      return {
        statusCode: statusCode,
        error: error,
        message: message
      };
    }
  }
};

function createHandler(handlerNames) {
  var handlerList = handlerNames.map(function (name) {
    if (!handlers[name]) {
      var availableHandlers = Object.keys(handlers).join(', ');
      throw new Error('No error handler named ' + name + '. Available are ' + availableHandlers);
    }
    return handlers[name];
  });

  handlerList.push(function (err) {
    return {
      statusCode: 500,
      message: err.message
    };
  });

  return function (err) {
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = handlerList[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var handler = _step.value;

        var res = handler(err);
        if (res !== undefined) {
          return res;
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  };
}

function errorHandler() {
  for (var _len = arguments.length, handlerNames = Array(_len), _key = 0; _key < _len; _key++) {
    handlerNames[_key] = arguments[_key];
  }

  return createHandler(handlerNames);
}

errorHandler.handlers = handlers;

exports.default = errorHandler;
module.exports = exports['default'];
