// routers/lib/handler-resolver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.default = function (root, opts) {
  var resolver = new HandlerResolver(root, opts);
  return resolver.resolve.bind(resolver);
};

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _es6require = require('@mattinsler/es6require');

var _es6require2 = _interopRequireDefault(_es6require);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function defaultParseSpecToDescriptor(spec) {
  var matches = spec.match(/^([^#]+)(#(.+))?$/);
  if (!matches) {
    throw new Error('Invalid handler format');
  }

  return {
    controller: matches[1],
    action: matches[3]
  };
}

function defaultDescriptorToHandler(_ref, _ref2) {
  var root = _ref.root;
  var controller = _ref2.controller;
  var action = _ref2.action;

  var handlerFile = require.resolve(_path2.default.join(root, controller));
  var handlerModule = (0, _es6require2.default)(handlerFile);
  if (!handlerModule[action]) {
    throw new Error('Could not find a method named ' + action + ' in ' + handlerFile);
  }

  return handlerModule[action];
}

var HandlerResolver = function () {
  function HandlerResolver(root) {
    var opts = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, HandlerResolver);

    this.root = _path2.default.resolve(process.cwd(), root || '');
    this.opts = {
      parseSpecToDescriptor: opts.parseSpecToDescriptor || defaultParseSpecToDescriptor,
      descriptorToHandler: opts.descriptorToHandler || defaultDescriptorToHandler
    };
  }

  _createClass(HandlerResolver, [{
    key: 'resolve',
    value: function resolve(spec) {
      if (typeof spec === 'string') {
        var _opts$parseSpecToDesc = this.opts.parseSpecToDescriptor(spec);

        var controller = _opts$parseSpecToDesc.controller;
        var _opts$parseSpecToDesc2 = _opts$parseSpecToDesc.action;
        var action = _opts$parseSpecToDesc2 === undefined ? 'index' : _opts$parseSpecToDesc2;

        return this.opts.descriptorToHandler(this, { controller: controller, action: action });
      }

      return spec;
    }
  }]);

  return HandlerResolver;
}();

module.exports = exports['default'];
