// routers/lib/routers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.promisedResolver = exports.handlerResolver = exports.classApiResolver = exports.apiResolver = exports.errorHandler = exports.guard = undefined;

var _guard2 = require('./guard');

var _guard3 = _interopRequireDefault(_guard2);

var _errorHandler2 = require('./error-handler');

var _errorHandler3 = _interopRequireDefault(_errorHandler2);

var _apiResolver2 = require('./api-resolver');

var _apiResolver3 = _interopRequireDefault(_apiResolver2);

var _classApiResolver2 = require('./class-api-resolver');

var _classApiResolver3 = _interopRequireDefault(_classApiResolver2);

var _handlerResolver2 = require('./handler-resolver');

var _handlerResolver3 = _interopRequireDefault(_handlerResolver2);

var _promisedResolver2 = require('./promised-resolver');

var _promisedResolver3 = _interopRequireDefault(_promisedResolver2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.guard = _guard3.default;
exports.errorHandler = _errorHandler3.default;
exports.apiResolver = _apiResolver3.default;
exports.classApiResolver = _classApiResolver3.default;
exports.handlerResolver = _handlerResolver3.default;
exports.promisedResolver = _promisedResolver3.default;
